rktl_planner/nodes/patrol_planner

```diff
@@ -0,0 +1,266 @@
+        self.LOOKAHEAD_DIST = rospy.get_param('~scoring/car_lookahead_dist', 1.0)
+        self.LOOKAHEAD_TIME = rospy.get_param('~scoring/ball_lookahead_time', 0.0)
+
+        # defense related parameters
+        self.DEFENSE_TIME_GAIN = rospy.get_param('~defense/reverse_time_gain', 0.5)
+
+        # variables
+        self.ball_position = None
+        self.ball_velocity = None
+        self.attempt_start_time = None
+
+        # Publishers
+        self.cmd_pub = rospy.Publisher('command', ControlCommand, queue_size=1)
+
+        # Subscribers
+        rospy.Subscriber('odom', Odometry, self.car_odom_cb)
+        rospy.Subscriber('/ball/odom', Odometry, self.ball_odom_cb)
+
+        rospy.spin()
+
+    def ball_odom_cb(self, odom_msg):
+        """Callback for ball odometry."""
+        self.ball_position = (
+            odom_msg.pose.pose.position.x,
+            odom_msg.pose.pose.position.y
+        )
+        self.ball_velocity = (
+            odom_msg.twist.twist.linear.x,
+            odom_msg.twist.twist.linear.y,
+        )
+
+    def car_odom_cb(self, odom_msg):
+        """Callback for car odometry."""
+
+        # extract car's position & heading
+        x = odom_msg.pose.pose.position.x
+        y = odom_msg.pose.pose.position.y
+        __, __, yaw = euler_from_quaternion([
+            odom_msg.pose.pose.orientation.x,
+            odom_msg.pose.pose.orientation.y,
+            odom_msg.pose.pose.orientation.z,
+            odom_msg.pose.pose.orientation.w
+        ])
+
+        # calculate distances / errors
+        NORTH_WALL_DISTANCE = abs(x - self.FIELD_HEIGHT/2.0)
+        SOUTH_WALL_DISTANCE = abs(x + self.FIELD_HEIGHT/2.0)
+        EAST_WALL_DISTANCE = abs(y + self.FIELD_WIDTH/2.0)
+        WEST_WALL_DISTANCE = abs(y - self.FIELD_WIDTH/2.0)
+
+        NORTH_HEADING_ERROR = sad(yaw, 0.0)
+        SOUTH_HEADING_ERROR = sad(yaw, math.pi)
+        EAST_HEADING_ERROR = sad(yaw, -math.pi/2.0)
+        WEST_HEADING_ERROR = sad(yaw, +math.pi/2.0)
+
+        # Prep command message for publish
+        cmd_msg = ControlCommand()
+
+        # check to see if an existing attempt has timed out
+        if self.attempt_start_time is not None and (rospy.Time.now() - self.attempt_start_time).to_sec() > self.ATTEMPT_TIMEOUT:
+            self.attempt_start_time = None
+            rospy.loginfo("attempt timeout")
+
+        # check that ball position is known
+        if self.ball_position is None or self.ball_velocity is None:
+            cmd_msg.velocity = 0.0
+            cmd_msg.curvature = 0.0
+            self.cmd_pub.publish(cmd_msg)
+            rospy.logwarn("ball position unknown")
+            return
+
+        # extract ball position and velocity, adding look-ahead time
+        bx, by = self.ball_position
+        vbx, vby = self.ball_velocity
+        bx += vbx * self.LOOKAHEAD_TIME
+        by += vby * self.LOOKAHEAD_TIME
+
+        # check if we should be offensive or defensive
+        if bx > 0.0:    # offensive
```

Comment:
Maybe we look into adding a slight offset. Like we may not want to be offensive until the ball is in the last 1/4th of the field.

Reply:
At least might be something to parametrize

Reply:
done